lightstep/b3_propagator.py

```diff
@@ -0,0 +1,136 @@
+                        "The value of x-b3-sampled should "
+                        "be {} instead of {}".format(
+                            int(sampled), sampled
+                        )
+                    )
+                carrier[_SAMPLED] = int(sampled)
+
+        if (sampled is None and flags is None) and (
+            traceid is None or spanid is None
+        ):
+            warn(
+                "If not propagating only the sampling state, traceid and "
+                "spanid must be defined"
+            )
+
+        carrier.update(baggage)
+
+        if traceid is not None:
+            carrier[_TRACEID] = format(traceid, "032x")
+        if spanid is not None:
+            carrier[_SPANID] = format(spanid, "016x")
+
+    def extract(self, carrier):
+
+        case_insensitive_carrier = {}
+        for key, value in carrier.items():
+            for b3_key in [
+                _SINGLE_HEADER,
+                _TRACEID,
+                _SPANID,
+                _PARENTSPANID,
+                _SAMPLED,
+                _FLAGS,
+            ]:
+                if re_compile(key, IGNORECASE).match(b3_key):
+                    case_insensitive_carrier[b3_key] = value
+                else:
+                    case_insensitive_carrier[key] = value
+
+        carrier = case_insensitive_carrier
+        baggage = {}
+
+        if _SINGLE_HEADER in carrier.keys():
+            fields = carrier[_SINGLE_HEADER].split("-", 4)
+            len_fields = len(fields)
+            if len(fields) == 1:
+                sampled = fields[0]
+            elif len(fields) == 2:
+                traceid, spanid = fields
+            elif len(fields) == 3:
+                traceid, spanid, sampled = fields
+            else:
+                traceid, spanid, sampled, parent_spanid = fields
+                baggage[_PARENTSPANID] = parent_spanid
+            if sampled == "d":
+                baggage[_FLAGS] = 1
+            else:
+                baggage[_SAMPLED] = sampled
+        else:
+            traceid = carrier.pop(_TRACEID, None)
+            spanid = carrier.pop(_SPANID, None)
+            parentspanid = carrier.pop(_PARENTSPANID, None)
+            sampled = carrier.pop(_SAMPLED, None)
+            flags = carrier.pop(_FLAGS, None)
+
+            if parentspanid is not None:
+                baggage[_PARENTSPANID] = parentspanid
+            if flags == 1:
+                baggage[_FLAGS] = flags
+                if sampled is not None:
+                    warn(
+                        "x-b3-flags: 1 implies x-b3-sampled: 1, ignoring "
+                        "the received value of x-b3-sampled"
+                    )
+            elif sampled is not None:
+                baggage[_SAMPLED] = sampled
+            baggage.update(carrier)
+            if baggage == OTSpanContext.EMPTY_BAGGAGE:
+                baggage = None
+        return SpanContext(
+            # traceid and spanid are encoded in hex, so thet must be encoded
```

Comment:
If no valid `traceid`/`spanid` were encoded, we need to report it as done here: https://github.com/opentracing/basictracer-python/blob/master/basictracer/text_propagator.py#L51

Reply:
Done, raising `SpanContextCorruptedException`.